### DISCUSSION TITLE
Flow Reading Accuracy Troubleshooting

### DISCUSSION THREAD

**Discussion by aneisch** [tronikos/esphome-magnetometer-water-gas-meter#19] [General]
First of all, thanks for all your work on this!

I got things set up with an ESP8266 (D1 Mini) and things initially looked perfect. However, during periods of higher flow (eg: irrigation) I found that my readings were considerably lower than expected and compared to my flume 2.

After some reading in issues in this repository I decided to rebuild my meter with an ESP32 (Lolin S2-mini). It seems to return slightly more accurate values at higher flow, but interestingly seems to also have a longer loop time (and reports "qmc5883l.sensor took a long time for an operation" ocasionally).

Has anyone experienced this, or do you have any suggestions for improving accuracy? I am using the same set up for each device with the screenshots below. I built a D1 mini shield with my QMC5883L that I simply moved between the two microcontrollers (leaving the sensor in the same position on the meter). For this test device I'm using ~1ft Cat5. 

D1 Mini Flow
![ESP8266 Flow](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/assets/6991461/04da6600-7658-47ea-83b8-eb0d2d970427)
D1 Mini Loop Time
![ESP8266](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/assets/6991461/67e0465b-3a5f-48f5-b7be-a72f57942a3d)
D1 Mini Config:
![Screen Shot 2024-07-01 at 5 38 24 PM](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/assets/6991461/6878b9f9-e8d4-4cac-b7d0-b604c3a0a599)

Compared to:

S2 Mini Flow
![S2-Mini Flow](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/assets/6991461/be96f4c1-5460-481a-8335-ade7cd5cff13)
S2 Mini Loop Time
![S2-Mini](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/assets/6991461/04d5dfd7-2e10-4477-94f0-8d8159d61569)
S2 Mini Configuration
![Screen Shot 2024-07-01 at 5 39 14 PM](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/assets/6991461/55faae03-d68b-4ab5-9a14-0ca3c276e25e)

**Comment by tronikos** [comment]
My recommendation is to try again without logging,  with the magnetic field sensors disabled, and maybe without the loop time sensor. If you have these enabled, at high flow the MCU has to push tons of data to Home Assistant which slow it down.

**Reply by aneisch** [reply]
I should have specified: I had the field sensors disabled in HA. Do I need to disable them in the yaml config as well or does disabling the entity in HA accomplish what we need?

I ordered a different esp32, hoping that I see an improvement. I will report back.

**Reply by tronikos** [reply]
I was referring to disabling them in HA. But since you mentioned, can you try converting them to internal in the config? Add this at the end of your config:

```yaml
sensor:
  - id: !extend ${prefix_id}qmc5883l_id
    field_strength_x:
      internal: true
    field_strength_y:
      internal: true
    field_strength_z:
      internal: true
```

**Comment by aneisch** [comment]
After switching from Lolin S2 Mini (single core) to Lolin S3 Mini (dual core) I am seeing sub 5ms loop times and improved accuracy, matching my Flume 2 during up to 25gal/m flows.

I also noted that regardless of whether the debug entities were enabled, I still saw a consistently higher loop time than if I have `internal: true` set for the axis sensors (typically 2x). I would suggest that others set this as well.

Full config:
```
substitutions:
  name: water_meter_s3
  friendly_name: "Water Meter S3"
  volume_unit: 'gal'
  i2c_scl: GPIO36  # D1
  i2c_sda: GPIO35  # D2
  prefix_name: s3
  flow_update_interval_seconds: "5"

esphome:
  name: ${name}
  platformio_options:
    board_build.arduino.memory_type: qio_qspi
    board_upload.ram_size: 327680
    board_upload_maximum_size: 4193404
    board_upload_speed: 460800
    build_flags: [
      "-DBOARD_HAS_PSRAM",
      "-DARDUINO_LOLIN_S3_MINI",
      "-DARDUINO_USB_MODE=1"
      ]

esp32:
  board: lolin_s3_mini
  variant: esp32s3
  flash_size: 4MB
  framework:
    type: arduino
    version: 2.0.8
    platform_version: 6.3.2

wifi:
  ssid: "SSID"
  password: !secret wifi_password

logger:
  level: DEBUG

api:
  #reboot_timeout: 0s # default
  encryption:
    key: !secret encryption_key

ota:
  platform: esphome
  password: esphome_recovery

i2c:
  id: i2c_bus
  sda: ${i2c_sda}
  scl: ${i2c_scl}
  scan: True
  frequency: 400kHz

packages:
  meter:
    url: https://github.com/tronikos/esphome-magnetometer-water-gas-meter
    ref: main
    file: esphome-water-meter.yaml
    # Or for gas meter:
    # file: esphome-gas-meter.yaml
    # Or if you are using HMC5883L instead of QMC5883L:
    # files: [esphome-water-meter.yaml, hmc5883l.yaml]
    refresh: 0s 

binary_sensor:
  - platform: template
    name: ${prefix_name}Flow Status
    lambda: |-
      if (id(${prefix_id}sensor_flow).state > 0) {
        return true;
      } else {
        return false;
      }
    
debug:
  update_interval: 5s

sensor:
  - platform: debug
    loop_time:
      name: "Loop Time"

  # loop time cut in half with internal: true
  - id: !extend ${prefix_id}qmc5883l_id
    field_strength_x:
      internal: true
    field_strength_y:
      internal: true
    field_strength_z:
      internal: true
```

**Reply by tronikos** [reply]
Great to hear you got it working! I'll update the README recommending a dual core ESP32.

It makes sense that `internal: true` is faster since ESPHome doesn't have to update Home Assistant. It seems when you disable an entity in Home Assistant, the integration keeps receiving updates from ESPHome, it just ignores them. I'll add a substitution to use `internal: true` by default and only expose them to HA if the user manually wants to calibrate. e.g. the calibrate button doesn't work well for them.

Try also adding this:
```yaml
sensor:
  - id: !extend ${prefix_id}sensor_half_rotations_total
    internal: true
```
This will save cycles continuously updating HA when water is flowing. Let me know if it's helpful so that I add a substitution for this too.

Separately of the above try adding the following:

```yaml
sensor:
  - id: !extend ${prefix_id}qmc5883l_id
    field_strength_x: !remove
    field_strength_y: !remove
    field_strength_z:
      internal: true
```
Make sure you adjust it first. The above assumes calibration picked the z axis.
This will save some more cycles updating the internal state of unused magnetic field sensors. If you see an improvement, I can add a recommendation in the README.

**Comment by tronikos** [comment]
I just committed:
1. [Set logger level to info to speed up device](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/commit/c2c5c2a8829f6b18ca7eec6bc68a4855157aa317)
2. [Mark some sensors internal to speed up device](https://github.com/tronikos/esphome-magnetometer-water-gas-meter/commit/bc316a1662f57829eb40277e43fe74cee0fdacd3)

On my ESP8266 there was a big improvement, especially with the 1st one. @aneisch make sure you remove the logger lines from your config. The loop time now remains constant around 10-12ms while water is running. Before it would jump to 50ms.

Removing field_strength_x and field_strength_y didn't have any noticeable impact for me.

@aneisch @jbronikowski could you please let me know what's the maximum flow you can measure with ESP8266, single core ESP32, and dual core ESP32 to update the README? All my faucets and shower heads are low flow.

**Reply by aneisch** [reply]
With latest changes (but hide half-rotations set to false as I didn't notice significant change):

S3 (dual core ESP32)): 25g/m 2-3ms avg loop time
S2 (single core ESP32): 25g/m 7-9ms avg loop time
ESP8266: 25g/m 4-5ms avg loop time

**Reply by tronikos** [reply]
What's your ESP8266? Getting 4-5ms is perfect. The fastest we can get data from the magnetometer is 5ms so any loop time less than that is perfect. Mine is a D1 mini and I get 10-12ms.

**Reply by aneisch** [reply]
Mine is a D1 mini pro. It looks like the clock speed is the same as a standard D1 mini though. Not sure how it's performing so well?

https://github.com/aneisch/home-assistant-config/blob/master/extras/esphome/water_meter8266.yaml

**Reply by tronikos** [reply]
I removed the api encryption and the loop time decreased by 2ms. Now it's 8-10ms. You might want to do the same as long as you understand the risks, i.e. any device in the same password protected WiFi network can access the ESP device. For me that's not a concern.
